From 798be43edd9ba5c11ebda70a1f5f4970167ca7af Mon Sep 17 00:00:00 2001
From: Kim Shepherd <kim@shepherd.nz>
Date: Tue, 1 Sep 2026 18:02:28 +0200
Subject: [PATCH] Hardening of OAI ingestion crosswalk

Use DSpaceHttpClient factory for ORE, disable redirects
Implement better allowed / forbidden host lists
Add integration tests

(cherry picked from commit 5a2a73ab48e86d804000811306ecd0e57569f09c)
---
 .../crosswalk/OREIngestionCrosswalk.java      | 286 +++++++++++-------
 .../crosswalk/OREIngestionCrosswalkTest.java  | 239 +++++++++++++++
 dspace/config/modules/oai.cfg                 |  27 +-
 3 files changed, 444 insertions(+), 108 deletions(-)
 create mode 100644 dspace-api/src/test/java/org/dspace/content/crosswalk/OREIngestionCrosswalkTest.java

diff --git a/dspace-api/src/main/java/org/dspace/content/crosswalk/OREIngestionCrosswalk.java b/dspace-api/src/main/java/org/dspace/content/crosswalk/OREIngestionCrosswalk.java
index 4401e0ea351c..08dac034ff66 100644
--- a/dspace-api/src/main/java/org/dspace/content/crosswalk/OREIngestionCrosswalk.java
+++ b/dspace-api/src/main/java/org/dspace/content/crosswalk/OREIngestionCrosswalk.java
@@ -10,21 +10,29 @@
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
-import java.net.ConnectException;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.net.URL;
 import java.sql.SQLException;
 import java.text.NumberFormat;
 import java.time.Instant;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Set;
 
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpResponse;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.logging.log4j.Logger;
+import org.dspace.app.client.DSpaceHttpClientFactory;
 import org.dspace.authorize.AuthorizeException;
 import org.dspace.content.Bitstream;
 import org.dspace.content.BitstreamFormat;
@@ -56,7 +64,7 @@
  * @author Alexey Maslov
  */
 public class OREIngestionCrosswalk
-    implements IngestionCrosswalk {
+        implements IngestionCrosswalk {
     /**
      * log4j category
      */
@@ -64,30 +72,30 @@ public class OREIngestionCrosswalk
 
     /* Namespaces */
     public static final Namespace ATOM_NS =
-        Namespace.getNamespace("atom", "http://www.w3.org/2005/Atom");
+            Namespace.getNamespace("atom", "http://www.w3.org/2005/Atom");
     private static final Namespace ORE_ATOM =
-        Namespace.getNamespace("oreatom", "http://www.openarchives.org/ore/atom/");
+            Namespace.getNamespace("oreatom", "http://www.openarchives.org/ore/atom/");
     private static final Namespace ORE_NS =
-        Namespace.getNamespace("ore", "http://www.openarchives.org/ore/terms/");
+            Namespace.getNamespace("ore", "http://www.openarchives.org/ore/terms/");
     private static final Namespace RDF_NS =
-        Namespace.getNamespace("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns#");
+            Namespace.getNamespace("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns#");
     private static final Namespace DCTERMS_NS =
-        Namespace.getNamespace("dcterms", "http://purl.org/dc/terms/");
+            Namespace.getNamespace("dcterms", "http://purl.org/dc/terms/");
     private static final Namespace DS_NS =
-        Namespace.getNamespace("ds", "http://www.dspace.org/objectModel/");
+            Namespace.getNamespace("ds", "http://www.dspace.org/objectModel/");
 
+    private String[] forbiddenHostUrls = null;
 
     protected BitstreamService bitstreamService = ContentServiceFactory.getInstance().getBitstreamService();
     protected BitstreamFormatService bitstreamFormatService = ContentServiceFactory.getInstance()
-                                                                                   .getBitstreamFormatService();
+            .getBitstreamFormatService();
     protected BundleService bundleService = ContentServiceFactory.getInstance().getBundleService();
     protected ItemService itemService = ContentServiceFactory.getInstance().getItemService();
     protected ConfigurationService configurationService = DSpaceServicesFactory.getInstance().getConfigurationService();
 
-
     @Override
     public void ingest(Context context, DSpaceObject dso, List<Element> metadata, boolean createMissingMetadataFields)
-        throws CrosswalkException, IOException, SQLException, AuthorizeException {
+            throws CrosswalkException, IOException, SQLException, AuthorizeException {
 
         // If this list contains only the root already, just pass it on
         if (metadata.size() == 1) {
@@ -104,7 +112,7 @@ public void ingest(Context context, DSpaceObject dso, List<Element> metadata, bo
 
     @Override
     public void ingest(Context context, DSpaceObject dso, Element root, boolean createMissingMetadataFields)
-        throws CrosswalkException, IOException, SQLException, AuthorizeException {
+            throws CrosswalkException, IOException, SQLException, AuthorizeException {
 
         Instant timeStart = Instant.now();
 
@@ -118,21 +126,26 @@ public void ingest(Context context, DSpaceObject dso, Element root, boolean crea
             return;
         }
 
+        var followRedirects = configurationService
+                .getBooleanProperty("oai.harvester.ore.follow-redirects", false);
+        var requestConfig = RequestConfig.custom()
+                .setRedirectsEnabled(followRedirects).build();
+
         Document doc = new Document();
         doc.addContent(root.detach());
 
         List<Element> aggregatedResources;
         String entryId;
         XPathExpression<Element> xpathLinks =
-            XPathFactory.instance()
+                XPathFactory.instance()
                         .compile("/atom:entry/atom:link[@rel=\"" + ORE_NS.getURI() + "aggregates" + "\"]",
-                                 Filters.element(), null, ATOM_NS);
+                                Filters.element(), null, ATOM_NS);
         aggregatedResources = xpathLinks.evaluate(doc);
 
         XPathExpression<Attribute> xpathAltHref =
-            XPathFactory.instance()
+                XPathFactory.instance()
                         .compile("/atom:entry/atom:link[@rel='alternate']/@href",
-                                 Filters.attribute(), null, ATOM_NS);
+                                Filters.attribute(), null, ATOM_NS);
         entryId = xpathAltHref.evaluateFirst(doc).getValue();
 
         // Next for each resource, create a bitstream
@@ -143,18 +156,24 @@ public void ingest(Context context, DSpaceObject dso, Element root, boolean crea
         for (Element resource : aggregatedResources) {
             String href = resource.getAttributeValue("href");
             log.debug("ORE processing: " + href);
+            String processedURL;
+            try {
+                processedURL = new URIBuilder(href).build().toString();
+            } catch (URISyntaxException e) {
+                throw new CrosswalkException("Could not parse URI: " + href, e);
+            }
 
             String bundleName;
             Element desc = null;
             XPathExpression<Element> xpathDesc =
-                XPathFactory.instance()
-                    .compile("/atom:entry/oreatom:triples/rdf:Description[@rdf:about=\"" +
-                                 this.encodeForURL(href) + "\"][1]",
-                             Filters.element(), null, ATOM_NS, ORE_ATOM, RDF_NS);
+                    XPathFactory.instance()
+                            .compile("/atom:entry/oreatom:triples/rdf:Description[@rdf:about=\"" +
+                                            this.encodeForURL(href) + "\"][1]",
+                                    Filters.element(), null, ATOM_NS, ORE_ATOM, RDF_NS);
             desc = xpathDesc.evaluateFirst(doc);
 
             if (desc != null && desc.getChild("type", RDF_NS).getAttributeValue("resource", RDF_NS)
-                                    .equals(DS_NS.getURI() + "DSpaceBitstream")) {
+                    .equals(DS_NS.getURI() + "DSpaceBitstream")) {
                 bundleName = desc.getChildText("description", DCTERMS_NS);
                 log.debug("Setting bundle name to: " + bundleName);
             } else {
@@ -167,63 +186,81 @@ public void ingest(Context context, DSpaceObject dso, Element root, boolean crea
             Bundle targetBundle;
 
             // if null, create the new bundle and add it in
-            if (targetBundles.size() == 0) {
+            if (targetBundles.isEmpty()) {
                 targetBundle = bundleService.create(context, item, bundleName);
                 itemService.addBundle(context, item, targetBundle);
             } else {
                 targetBundle = targetBundles.get(0);
             }
 
-            URL ARurl = null;
-            InputStream in = null;
-            if (href != null) {
-                try {
-                    // Make sure the url string escapes all the oddball characters
-                    String processedURL = encodeForURL(href);
-                    if (validResourceUri(entryId, processedURL)) {
-                        // Generate a request for the aggregated resource
-                        ARurl = new URL(processedURL);
-                        in = ARurl.openStream();
-                    } else {
-                        throw new FileNotFoundException("Failed to validate " + processedURL);
-                    }
-                } catch (FileNotFoundException fe) {
-                    log.error("The provided URI failed to return a resource: " + href);
-                } catch (ConnectException fe) {
-                    log.error("The provided URI was invalid: " + href);
-                }
-            } else {
-                throw new CrosswalkException("Entry did not contain link to resource: " + entryId);
+            if (StringUtils.isBlank(href)) {
+                throw new CrosswalkException("The href attribute is required for the ORE ingestion.");
             }
-
-            // ingest and update
-            if (in != null) {
-                Bitstream newBitstream = bitstreamService.create(context, targetBundle, in);
-
-                String bsName = resource.getAttributeValue("title");
-                newBitstream.setName(context, bsName);
-
-                // Identify the format
-                String mimeString = resource.getAttributeValue("type");
-                BitstreamFormat bsFormat = bitstreamFormatService.findByMIMEType(context, mimeString);
-                if (bsFormat == null) {
-                    bsFormat = bitstreamFormatService.guessFormat(context, newBitstream);
+            try (CloseableHttpClient httpClient = DSpaceHttpClientFactory.getInstance()
+                    .buildWithRequestConfig(requestConfig)) {
+                if (!validResourceUri(processedURL)) {
+                    throw new FileNotFoundException("Invalid resource URI: " + processedURL);
+                }
+                // Generate a request for the aggregated resource
+                HttpGet httpGet = new HttpGet(processedURL);
+                HttpResponse response = httpClient.execute(httpGet);
+                if (response == null || response.getEntity() == null) {
+                    throw new FileNotFoundException(processedURL + " returned a null response or body");
+                }
+                if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+                    throw new FileNotFoundException(processedURL
+                            + " returned a " + response.getStatusLine() + " response");
+                }
+                if (response.getEntity() == null || response.getEntity().getContent() == null) {
+                    throw new FileNotFoundException(processedURL + " returned an empty body");
+                }
+                // ingest and update
+                try (InputStream in = response.getEntity().getContent()) {
+                    ingestStreamAsBitstream(context, in, targetBundle, resource, entryId);
                 }
-                newBitstream.setFormat(context, bsFormat);
-                bitstreamService.update(context, newBitstream);
-
-                bundleService.addBitstream(context, targetBundle, newBitstream);
-                bundleService.update(context, targetBundle);
-            } else {
-                throw new CrosswalkException("Could not retrieve bitstream: " + entryId);
             }
-
         }
         log.info(
-            "OREIngest for Item " + item.getID() + " took: " +
-                (Instant.now().toEpochMilli() - timeStart.toEpochMilli()) + "ms.");
+                "OREIngest for Item " + item.getID() + " took: " +
+                        (Instant.now().toEpochMilli() - timeStart.toEpochMilli()) + "ms.");
     }
 
+    /**
+     * Read an input stream and ingest it as a bitstream to the target bundle
+     * @param context Dspace context
+     * @param inputStream input stream containing bitstream content
+     * @param targetBundle the target bundle for the new bitstream
+     * @param resource the ORE resource Element
+     * @param entryId the entry ID of the ORE resource
+     * @throws AuthorizeException if current user does not have permission to add / create bitstream
+     * @throws IOException if the input stream is null or unreadable
+     * @throws SQLException if bitstream or bundle database operations fail
+     */
+    void ingestStreamAsBitstream(Context context, InputStream inputStream,
+                                         Bundle targetBundle, Element resource, String entryId)
+            throws AuthorizeException, IOException, SQLException {
+        // ingest and update
+        if (inputStream != null) {
+            Bitstream newBitstream = bitstreamService.create(context, targetBundle, inputStream);
+
+            String bsName = resource.getAttributeValue("title");
+            newBitstream.setName(context, bsName);
+
+            // Identify the format
+            String mimeString = resource.getAttributeValue("type");
+            BitstreamFormat bsFormat = bitstreamFormatService.findByMIMEType(context, mimeString);
+            if (bsFormat == null) {
+                bsFormat = bitstreamFormatService.guessFormat(context, newBitstream);
+            }
+            newBitstream.setFormat(context, bsFormat);
+            bitstreamService.update(context, newBitstream);
+
+            bundleService.addBitstream(context, targetBundle, newBitstream);
+            bundleService.update(context, targetBundle);
+        } else {
+            throw new IOException("Could not read input stream for: " + entryId);
+        }
+    }
 
     /**
      * Helper method to escape all characters that are not part of the canon set
@@ -265,58 +302,101 @@ private String encodeForURL(String sourceString) {
 
     /**
      * Validate a resource URI against the host and scheme of the remote OAI endpoint, or a configured
-     * list of allowed prefixes.
-     * This still implicitly "trusts" the remote OAI server, but will reject resource URIs with a totally
-     * different hostname to avoid downloading malicious resources from a compromised endpoint.
+     * list of allowed prefixes. Some default forbidden hosts are also included to prevent SSRF.
      * Even if the URL prefix validation is disabled, schemes will still be enforced to http(s) so file:/// and
      * other unwanted schemes cannot be used
-     * @param entryUrl the entryId of the parent ORE resource
      * @param resourceUrl the resource URL of the aggregated ORE resource
      * @return result of the validation
      */
-    private boolean validResourceUri(String entryUrl, String resourceUrl) {
+    boolean validResourceUri(String resourceUrl) {
+        URI resourceUri;
         try {
-            Set<String> allowedSchemes = Set.of("http", "https");
-            URI entryUri = new URI(entryUrl).normalize();
-            URI resourceUri = new URI(resourceUrl).normalize();
-            String scheme = resourceUri.getScheme();
-
-            if (scheme == null ||
-                    !allowedSchemes.contains(scheme.toLowerCase(Locale.ROOT))) {
-                log.warn("Illegal scheme requested for ORE resource: {}", resourceUri);
-                return false;
-            }
+            resourceUri = new URI(resourceUrl).normalize();
+        } catch (URISyntaxException | NullPointerException e) {
+            log.error("Invalid resource URI: " + resourceUrl);
+            return false;
+        }
+        String resourceHost = resourceUri.getHost();
 
-            if (configurationService.getBooleanProperty("oai.harvester.ore.file.validateUrlPrefix", false)) {
-                for (String allowedPrefix : configurationService
-                        .getArrayProperty("oai.harvester.ore.file.allowedUrlPrefix")) {
-                    URI allowedUri = new URI(allowedPrefix).normalize();
-                    // Return true on the first allowed prefix match
-                    if (Objects.equals(resourceUri.getScheme(), allowedUri.getScheme())
-                            && Objects.equals(resourceUri.getHost().toLowerCase(Locale.ROOT),
-                            allowedUri.getHost().toLowerCase(Locale.ROOT))) {
-                        return true;
-                    }
-                }
+        Set<String> allowedSchemes = Set.of("http", "https");
+        String resourceScheme = resourceUri.getScheme();
+        if (resourceScheme == null ||
+                !allowedSchemes.contains(resourceScheme.toLowerCase(Locale.ROOT))) {
+            log.warn("Illegal scheme requested for ORE resource: {}", resourceUrl);
+            return false;
+        }
 
-                // If no allowed prefixes were matched, we require scheme + host to match the remote OAI server
-                if (!Objects.equals(entryUri.getScheme(), resourceUri.getScheme())) {
-                    log.warn("Illegal scheme requested for ORE resource: {}", resourceUri);
-                    return false;
-                }
-                if (!Objects.equals(
-                        entryUri.getHost().toLowerCase(Locale.ROOT),
-                        resourceUri.getHost().toLowerCase(Locale.ROOT))) {
-                    log.warn("Illegal host requested for ORE resource: {}", resourceUri);
+        initializeForbiddenHosts();
+        if (forbiddenHostUrls != null) {
+            for (String forbiddenHostUrl : forbiddenHostUrls) {
+                String forbiddenHost = extractHostname(forbiddenHostUrl);
+                if (forbiddenHost != null && Objects.equals(forbiddenHost, resourceHost)) {
+                    log.warn("Forbidden hostname in ORE resource URL: {}", resourceUrl);
                     return false;
                 }
             }
+        }
 
+        // This is now very strict - it is up to administrators to configure a list of allowed URL prefixes.
+        // If validateHost is set to 'true', and an empty list will cause every resource URL to be rejected
+        // Unlike forbidden URLs, we expect these to be hostnames rather than partial or full URIs as well
+        if (configurationService.getBooleanProperty("oai.harvester.ore.file.validateHost", false)) {
+            for (String allowedHost : configurationService
+                    .getArrayProperty("oai.harvester.ore.file.allowedHosts")) {
+                if (Objects.equals(resourceUri.getHost().toLowerCase(Locale.ROOT),
+                        allowedHost.toLowerCase(Locale.ROOT))) {
+                    return true;
+                }
+            }
+        } else {
             return true;
+        }
+        return false;
+    }
 
+    /**
+     * To keep config easier to maintain with both plain hostnames like 'bad.example.com' and references
+     * to partial URLs elsewhere in configuration like ${dspace.ui.url}, we test to see which it is
+     * and return the string representing the hostname.
+     * @param urlOrHost URL, partial URL or hostname to extract and validate
+     * @return extracted hostname or null
+     */
+    String extractHostname(String urlOrHost) {
+        if (StringUtils.isBlank(urlOrHost)) {
+            return null;
+        }
+        try {
+            URI uri = new URI(urlOrHost);
+            String host = uri.getHost();
+            if (host != null) {
+                return host.toLowerCase(Locale.ROOT);
+            }
         } catch (URISyntaxException e) {
-            log.warn("Could not validate ORE resource URI: {}", resourceUrl);
-            return false;
+            // pass through
+        }
+        // Return plain lower-cased string as defaulg
+        return urlOrHost.toLowerCase(Locale.ROOT);
+    }
+
+    /**
+     * Initialize forbidden hosts from config, if not set already
+     */
+    void initializeForbiddenHosts() {
+        if (forbiddenHostUrls == null) {
+            String dspaceServerUrl = configurationService.getProperty("dspace.server.url");
+            String dspaceUiUrl = configurationService.getProperty("dspace.ui.url");
+            String[] localAndEC2Hosts = {"localhost", "127.0.0.1", "[::1]", "169.254.169.254", "[fd00:ec2::254]"};
+            List<String> defaultForbiddenUrlPrefixes = new ArrayList<>();
+            Collections.addAll(defaultForbiddenUrlPrefixes, localAndEC2Hosts);
+            if (StringUtils.isNotBlank(dspaceServerUrl)) {
+                defaultForbiddenUrlPrefixes.add(dspaceServerUrl);
+            }
+            if (StringUtils.isNotBlank(dspaceUiUrl)) {
+                defaultForbiddenUrlPrefixes.add(dspaceUiUrl);
+            }
+            forbiddenHostUrls = configurationService.getArrayProperty(
+                    "oai.harvester.ore.file.forbiddenHostUrls",
+                    defaultForbiddenUrlPrefixes.toArray(new String[0]));
         }
     }
 
diff --git a/dspace-api/src/test/java/org/dspace/content/crosswalk/OREIngestionCrosswalkTest.java b/dspace-api/src/test/java/org/dspace/content/crosswalk/OREIngestionCrosswalkTest.java
new file mode 100644
index 000000000000..c4662805ac06
--- /dev/null
+++ b/dspace-api/src/test/java/org/dspace/content/crosswalk/OREIngestionCrosswalkTest.java
@@ -0,0 +1,239 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.content.crosswalk;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.dspace.AbstractIntegrationTestWithDatabase;
+import org.dspace.app.client.DSpaceHttpClientFactory;
+import org.dspace.builder.BundleBuilder;
+import org.dspace.builder.CollectionBuilder;
+import org.dspace.builder.CommunityBuilder;
+import org.dspace.builder.ItemBuilder;
+import org.dspace.content.Bundle;
+import org.dspace.content.Collection;
+import org.dspace.content.Item;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.ItemService;
+import org.dspace.eperson.EPerson;
+import org.jdom2.Element;
+import org.jdom2.Namespace;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+/**
+ * Test class for OREIngestionCrosswalk
+ *
+ * @author Kim Shepherd
+ */
+public class OREIngestionCrosswalkTest extends AbstractIntegrationTestWithDatabase {
+
+    private Bundle bundle;
+    private OREIngestionCrosswalk crosswalk;
+    private Item item;
+    private ItemService itemService;
+    private EPerson prevUser;
+
+    private static final Namespace ATOM_NS = Namespace.getNamespace("atom", "http://www.w3.org/2005/Atom");
+    private static final Namespace ORE_NS = Namespace.getNamespace("ore", "http://www.openarchives.org/ore/terms/");
+
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        crosswalk = new OREIngestionCrosswalk();
+        itemService = ContentServiceFactory.getInstance().getItemService();
+
+        context.turnOffAuthorisationSystem();
+        parentCommunity = CommunityBuilder.createCommunity(context).build();
+        Collection collection = CollectionBuilder.createCollection(context, parentCommunity).build();
+        item = ItemBuilder.createItem(context, collection).withTitle("Test ORE Object").build();
+        bundle = BundleBuilder.createBundle(context, item).withName("ORIGINAL").build();
+        context.restoreAuthSystemState();
+        prevUser = context.getCurrentUser();
+        context.setCurrentUser(admin);
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        context.setCurrentUser(prevUser);
+    }
+
+    @Test
+    public void testIngestNullRoot() throws Exception {
+        // should just return without exception
+        crosswalk.ingest(context, item, (Element) null, false);
+    }
+
+    @Test
+    public void testIngestEmptyResources() throws Exception {
+        Element entry = createOREEntryWithoutResources();
+        crosswalk.ingest(context, item, entry, false);
+
+        List<Bundle> bundles = itemService.getBundles(item, "ORIGINAL");
+        assertTrue("Should have no bundles or empty bundle",
+                bundles.isEmpty() || bundles.get(0).getBitstreams().isEmpty());
+    }
+
+    @Test(expected = CrosswalkException.class)
+    public void testIngestInvalidURISyntaxThrowsException() throws Exception {
+        Element entry = createOREEntryWithInvalidURI();
+        crosswalk.ingest(context, item, entry, false);
+    }
+
+    @Test
+    public void testHostValidation() {
+        // localhost = forbidden
+        boolean result = crosswalk.validResourceUri("http://localhost:8080/resource");
+        assertFalse("localhost should be forbidden", result);
+
+        // 127.0.0.1 = forbidden
+        result = crosswalk.validResourceUri("http://127.0.0.1:8080/resource");
+        assertFalse("127.0.0.1 should be forbidden", result);
+
+        // allowed.example.com = not forbidden
+        result = crosswalk.validResourceUri("http://allowed.example.com/resource");
+        assertTrue("External host should be allowed", result);
+    }
+
+    @Test
+    public void testSchemeValidation() throws Exception {
+        boolean result = crosswalk.validResourceUri("file:///etc/passwd");
+        assertFalse("file:// scheme should be forbidden", result);
+
+        result = crosswalk.validResourceUri("ftp://example.com/resource");
+        assertFalse("ftp:// scheme should be forbidden", result);
+
+        result = crosswalk.validResourceUri("https://example.com/resource");
+        assertTrue("https:// scheme should be allowed", result);
+    }
+
+    @Test
+    public void testSuccessfulBitstreamIngest() throws Exception {
+        // fully mocked response
+        try (MockedStatic<DSpaceHttpClientFactory> mockedFactory =
+                     mockHttpClient(200, "Test bitstream content")) {
+
+            Element entry = createValidOREEntry();
+            crosswalk.ingest(context, item, entry, false);
+
+            List<Bundle> bundles = itemService.getBundles(item, "ORIGINAL");
+            assertFalse("Should have created bundle", bundles.isEmpty());
+            assertFalse("Bundle should have bitstream", bundles.get(0).getBitstreams().isEmpty());
+        }
+    }
+
+    @Test
+    public void testIngestStreamAsBitstream() throws Exception {
+        String testContent = "Test bitstream content";
+        InputStream inputStream = new ByteArrayInputStream(testContent.getBytes());
+
+        Element resource = new Element("link", ATOM_NS);
+        resource.setAttribute("href", "http://example.com/test.pdf");
+        resource.setAttribute("title", "Test Document");
+        resource.setAttribute("type", "application/pdf");
+
+        crosswalk.ingestStreamAsBitstream(context, inputStream, bundle, resource, "test-entry");
+
+        assertFalse("Bundle should have bitstream", bundle.getBitstreams().isEmpty());
+        assertEquals("Bitstream should have correct name", "Test Document",
+                bundle.getBitstreams().get(0).getName());
+    }
+
+    @Test(expected = IOException.class)
+    public void testIngestStreamAsBitstreamNullStreamThrowsException() throws Exception {
+        Element resource = new Element("link", ATOM_NS);
+        resource.setAttribute("href", "http://example.com/test.pdf");
+
+        crosswalk.ingestStreamAsBitstream(context, null, bundle, resource, "test-entry");
+    }
+
+    private Element createValidOREEntry() {
+        Element entry = new Element("entry", ATOM_NS);
+        Element altLink = new Element("link", ATOM_NS);
+        altLink.setAttribute("rel", "alternate");
+        altLink.setAttribute("href", "http://example.com/entry/123");
+        entry.addContent(altLink);
+        Element resLink = new Element("link", ATOM_NS);
+        resLink.setAttribute("rel", ORE_NS.getURI() + "aggregates");
+        resLink.setAttribute("href", "http://example.com/resource.pdf");
+        resLink.setAttribute("title", "Test Resource");
+        resLink.setAttribute("type", "application/pdf");
+        entry.addContent(resLink);
+        return entry;
+    }
+
+    private Element createOREEntryWithoutResources() {
+        Element entry = new Element("entry", ATOM_NS);
+        Element altLink = new Element("link", ATOM_NS);
+        altLink.setAttribute("rel", "alternate");
+        altLink.setAttribute("href", "http://example.com/entry/123");
+        entry.addContent(altLink);
+        return entry;
+    }
+
+
+    private Element createOREEntryWithInvalidURI() {
+        Element entry = new Element("entry", ATOM_NS);
+        Element altLink = new Element("link", ATOM_NS);
+        altLink.setAttribute("rel", "alternate");
+        altLink.setAttribute("href", "http://example.com/entry/123");
+        entry.addContent(altLink);
+        Element resLink = new Element("link", ATOM_NS);
+        resLink.setAttribute("rel", ORE_NS.getURI() + "aggregates");
+        resLink.setAttribute("href", "hello, i am an invalid URI");
+        entry.addContent(resLink);
+        return entry;
+    }
+
+    private MockedStatic<DSpaceHttpClientFactory> mockHttpClient(int statusCode, String content) throws IOException {
+        MockedStatic<DSpaceHttpClientFactory> mockedFactory = Mockito.mockStatic(DSpaceHttpClientFactory.class);
+
+        CloseableHttpClient mockClient = mock(CloseableHttpClient.class);
+        DSpaceHttpClientFactory mockFactoryInstance = mock(DSpaceHttpClientFactory.class);
+        CloseableHttpResponse mockResponse = mockHttpResponse(statusCode, content);
+
+        when(mockFactoryInstance.buildWithRequestConfig(any())).thenReturn(mockClient);
+        when(mockClient.execute(any(HttpGet.class))).thenReturn(mockResponse);
+        mockedFactory.when(DSpaceHttpClientFactory::getInstance).thenReturn(mockFactoryInstance);
+
+        return mockedFactory;
+    }
+
+    private CloseableHttpResponse mockHttpResponse(int statusCode, String content) throws IOException {
+        CloseableHttpResponse mockResponse = mock(CloseableHttpResponse.class);
+        HttpEntity mockEntity = mock(HttpEntity.class);
+        StatusLine mockStatus = mock(StatusLine.class);
+
+        when(mockStatus.getStatusCode()).thenReturn(statusCode);
+        when(mockEntity.getContent()).thenReturn(new ByteArrayInputStream(content.getBytes()));
+        when(mockResponse.getStatusLine()).thenReturn(mockStatus);
+        when(mockResponse.getEntity()).thenReturn(mockEntity);
+
+        return mockResponse;
+    }
+
+}
